lib/puppetserver/ca/create_action.rb

```diff
@@ -0,0 +1,253 @@
+          save(new_public_key, certname, settings[:publickeydir])
+          @logger.inform "Keys for #{certname} saved in #{settings[:privatekeydir]} and #{settings[:publickeydir]}"
+          new_csr = host.create_csr(certname, new_private_key)
+          certname_csr_hash[certname] = new_csr.to_s
+        end
+        certname_csr_hash
+      end
+
+      def save(content, certname, dir)
+        location = File.join(dir, "#{certname}.pem")
+        @logger.warn "#{location} already exists, overwriting" if File.exist?(location)
+        Puppetserver::Utils::FileUtilities.write_file(location, content, 0640)
+      end
+
+
+      # Make an HTTP request to submit certificate requests to CA
+      # @param certname_csr_hash [Hash] hash with certname keys and csr_string values
+      # @param csr [String] string version of a OpenSSL::X509::Request
+      # @param settings [Hash] a hash of config settings
+      # @return [Boolean] success of all csrs being submitted to CA
+      def submit_certificate_requests(certname_csr_hash, settings)
+        url = HttpUtilities.make_ca_url(settings[:ca_server],
+                       settings[:ca_port],
+                       'certificate_request')
+
+        HttpUtilities.with_connection(url, settings) do |connection|
+          certname_csr_hash.each do |certname, csr_string|
+            url.resource_name = certname
+            result = connection.put(csr_string, url)
+
+            log_csr_result(result, certname)
+          end
+        end
+      end
+
+      def log_csr_result(result, certname)
+        case result.code
+        when '200', '204'
+          @logger.inform "#{certname} has a waiting certificate request"
+        else
+          @logger.err 'Error:'
+          @logger.err "    When request submitted for #{certname}:"
+          @logger.err "      code: #{result.code}"
+          @logger.err "      body: #{result.body.to_s}" if result.body
+        end
+      end
+
+      # Make an HTTP request to CA to sign the named certificates
+      # @param certnames [Array] the names of the certificates to have signed
+      # @return [Boolean] the success of all certificates being signed
+      def sign_certs(certnames, settings)
+        url = HttpUtilities.make_ca_url(settings[:ca_server],
+                       settings[:ca_port],
+                       'certificate_status')
+
+        HttpUtilities.with_connection(url, settings) do |connection|
+          certnames.each do |certname|
+            url.resource_name = certname
+            body = JSON.dump({desired_state: 'signed'})
+            result = connection.put(body, url)
+            log_sign_result(result, certname)
+          end
+        end
+      end
+
+      def log_sign_result(result, certname)
+        case result.code
+        when '204'
+          @logger.inform "Signed certificate request for #{certname}"
+        else
+          @logger.err 'Error:'
+          @logger.err "    When download requested for #{certname}:"
```

Comment:
Oh, I like this error message; it's formatted a lot better than mine and gives more valuable context. I might need to steal it.

Reply:
^ I stole it from Justin :)